### PR TITLE
Add left recursion check: quit early instead of going into an infinite loop

### LINKED ISSUE #6492: Segfault when parsing grammars with left-recursion
Using latest master branch on OSX Sonoma 14.2.1 (23C71), M1 Macbook Pro.

When parsing grammars that contain left-recursion, the application parses with a segfault.

Example:
```
./main -m ./models/llama-2-7b.Q4_0.gguf --grammar "root ::= root \"a\" | \"b\""
```
Results in:
```
system_info: n_threads = 8 / 10 | AVX = 0 | AVX_VNNI = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | SSSE3 = 0 | VSX = 0 | MATMUL_INT8 = 0 |
sampling:
	repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
	top_k = 40, tfs_z = 1.000, top_p = 0.950, min_p = 0.050, typical_p = 1.000, temp = 0.800
	mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000
sampling order:
CFG -> Penalties -> top_k -> tfs_z -> typical_p -> top_p -> min_p -> temperature
generate: n_ctx = 512, n_batch = 2048, n_predict = -1, n_keep = 1


[1]    50441 segmentation fault  ./main -m ./models/llama-2-7b.Q4_0.gguf --grammar
```

The segfault occurs in `llama_grammar_init` -- still digging into it, but wanted to at least log the bug.

This is not a regression, and is different from previous grammar-related segfault errors in the past.

This probably affects the server, and represents a vector to crash the server with a malformed request.

Ideally we should be able to catch left-recursion situations and hopefully either continue gracefully, or at the very least fail with an intelligent error message.

### PR DESCRIPTION
**Pull request by nuchi**
Fixes https://github.com/ggerganov/llama.cpp/issues/6492

Adds a check to see if the grammar has left recursion, and quits instead of going into an infinite loop.

I'm not sure if this is the best place for it, happy to move it around if not.

attn: @HanClinto

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -13085,6 +13085,25 @@ static std::vector<llama_grammar_candidate> llama_grammar_reject_candidates(
...
+    LLAMA_LEFT_REC_IN_PROGRESS     = 1,
+
+    // finished searching this nonterminal
+    LLAMA_LEFT_REC_FINISHED_SEARCH = 2,
+
+    // detected a cycle
+    LLAMA_LEFT_REC_FOUND_CYCLE     = 3,
+};
+
+static void detect_left_recursion(
```
Comment (HanClinto) [inline llama.cpp:13102]: Minor formatting note -- internal functions and structures such as these should go in the section marked `grammar - internal`.

Also, because this is a grammar-specific function, it should probably be renamed `llama_grammar_detect_left_recursion`

**Comment by HanClinto** [comment]
Nice work, @nuchi !

Very sorry for missing this when you first opened it -- I have adjusted my notification settings and hopefully I won't miss such things in the future.

This looks really good. I especially like your extensions of the integration tests -- well done!

Reading your code, I was initially worried that it might falsely flag examples of non-left recursion, but you handled that perfectly.

I like your approach of checking for left recursion as part of the grammar initialization. My first approach was to detect left recursion only at inference time (inside of llama_grammar_advance_stack), but that is non-deterministic -- because if a user has left recursion buried deep in their grammar, it's possible one could generate text that doesn't hit it -- but your check is exhaustive, which I think is better than how I was going to approach it.

Looking at the code, I think it can be simplified quite a bit. For instance, you're doing a full walk through the tree every time, but there is possibility to have an early exit in the case of finding recursion.

Also, the custom enumeration is cool (and very thorough!), but might be a bit overkill for our needs. All we need to do is track lists of the rules references that are visited in each branch, and check to see if they have been visited before. I took your base code and stripped it down a bit:

```
static bool llama_grammar_detect_left_recursion(
        const std::vector<std::vector<llama_grammar_element>> & rules,
        size_t                                                  rule_index,
        std::vector<size_t>                                   * rules_visited) {

    if (rules_visited == nullptr) {
        rules_visited = new std::vector<size_t>();
    }

    // Check to see if this rule index has already been visited
    if (std::find(rules_visited->begin(), rules_visited->end(), rule_index) != rules_visited->end()) {
        // If it has, then we're in a loop
        return true;
    }

    // Mark that we have looked at this rule
    rules_visited->emplace_back(rule_index);

    // Recurse through this rule's elements
    const std::vector<llama_grammar_element> & rule = rules[rule_index];
    size_t i = 0;
    do {
        // If the first element in a rule sequence is a rule reference, recurse into that rule
        if (rule[i].type == LLAMA_GRETYPE_RULE_REF) {
            // Create a new vector to store the list of rules visited in this branch
            std::vector<size_t> rules_visited_branch(*rules_visited);
            if (llama_grammar_detect_left_recursion(rules, (size_t)rule[i].value, &rules_visited_branch)) {
                // Early exit if we detect left recursion
                return true;
            }
        }
        // Skip over all other elements in this sequence -- even if they are rule references.
        // We only care about detecting left recursion -- not all recursion.
        while (!llama_grammar_is_end_of_sequence(&rule[i])) {
            i++;
        }
        i++;
    } while (i < rule.size());

    return false;
}
```

And then this would be called from `llama_grammar_init` in this way:

```
    // Check for left recursion
    for (size_t i = 0; i < n_rules; i++) {
        if (llama_grammar_detect_left_recursion(vec_rules, i, nullptr)) {
            throw std::runtime_error(format("unsupported grammar, left recursion detected for nonterminal at index %d", (int)i));
        }
    }
```

This is not new code -- I only adapted what you already had here and tried to reduce it to your bare essentials.

Overall I love your approach, and it's _much_ better than the direction I was originally taking. Well done!

File: tests/test-grammar-integration.cpp [hunk 2]
```diff
@@ -28,6 +28,18 @@ static llama_grammar* build_grammar(const std::string & grammar_str) {
...
+    try {
+        build_grammar(grammar_str);
+        fprintf(stderr, "❌ Expected build failure, but succeeded: %s\n", grammar_str.c_str());
+    } catch (const std::exception & err) {
+        grammar_fails = true;
+        fprintf(stdout, "✅︎\n");
+    }
+    return grammar_fails;
+}
+
```
Comment (HanClinto) [inline tests/test-grammar-integration.cpp:42]: Minor formatting adjustment so that we print the grammar on success as well.
```suggestion
static bool test_build_grammar_fails(const std::string & grammar_str) {
    fprintf(stderr, "⚫ Testing failure for grammar: %s\n", grammar_str.c_str());

    bool grammar_fails = false;
    try {
        build_grammar(grammar_str);
        fprintf(stderr, "  ❌ Expected build failure, but succeeded\n");
    } catch (const std::exception & err) {
        grammar_fails = true;
        fprintf(stdout, "  ✅︎\n");
    }
    return grammar_fails;
}
```

**Comment by HanClinto** [comment]
I especially like how you extended the integration tests here. Super well done!

**Comment by nuchi** [comment]
Great suggestions, thank you, good idea to ditch the enum. I'll incorporate them with maybe some minor changes — I also just realized I missed an edge case where the leftmost nonterminal(s) might be empty (in which case I also need to recurse into the next-leftmost nonterminal).

> Very sorry for missing this when you first opened it -- I have adjusted my notification settings and hopefully I won't miss such things in the future.

Sorry about that and probably my fault — I edited your @ into my initial post, so maybe that's why it didn't ping you.

Finally I bring up that it's also a possibility to return nullptr and leave handling to the caller. That'd be messy to do right now since callers aren't expecting it, but also it might be preferable in the case of e.g. a server accepting grammars over the wire where we don't want the server to fall over on a bad request.

**Comment by HanClinto** [comment]
> I also just realized I missed an edge case where the leftmost nonterminal(s) might be empty (in which case I also need to recurse into the next-leftmost nonterminal).

Oh nice! Were you able to build this into another test case?

> > Very sorry for missing this when you first opened it -- I have adjusted my notification settings and hopefully I won't miss such things in the future.
> 
> Sorry about that and probably my fault — I edited your @ into my initial post, so maybe that's why it didn't ping you.

That might have contributed -- I was also being e-mailed on every PR, which is cool sometimes, but it also added to a bit too much noise, so now I'm back down to participating and mentions. :)

> Finally I bring up that it's also a possibility to return nullptr and leave handling to the caller. That'd be messy to do right now since callers aren't expecting it, but also it might be preferable in the case of e.g. a server accepting grammars over the wire where we don't want the server to fall over on a bad request.

I agree -- that feels cleaner. 

Along those same lines, you might consider moving this into `llama_sampling_init()` (in sampling.cpp) where other grammar correctness checks are made -- I.E., right next to the root node check is made. Checking for left recursion could be right after that, and return a nullptr in the failing case. Then all of the validation checks are made in the same place.

Only awkward bit with that implementation is that you wouldn't have the rules split up already into those nice vectors, so it might be a bit more pointer math at that stage in the processing, but it's something to consider.

**Comment by nuchi** [comment]
Pushed an update but I still intend to take a look at the interface and maybe return nullptr. I may also look at removing some code repetition.

**Comment by nuchi** [comment]
Actually, I might leave it as is for now if that looks good to you. I have some ideas about interface changes but I think that'd be a larger change that's decoupled from this one. As of now, the changes here are self-contained and strictly improvements to the status quo, in that the thrown exception here is what would have been a stack overflow or OOM in the status quo.

File: llama.cpp
Comment (HanClinto) [inline llama.cpp:1]: Don't need this definition here, since the function is defined right below it.

File: llama.cpp [hunk 3]
```diff
@@ -13174,28 +13174,68 @@ static std::vector<llama_grammar_candidate> llama_grammar_reject_candidates(
...
 
-//
-// grammar - external
-//
+static bool llama_grammar_detect_left_recursion(
+        const std::vector<std::vector<llama_grammar_element>> & rules,
+        size_t                                                  rule_index,
+        std::vector<bool>                                     * rules_visited,
+        std::vector<bool>                                     * rules_in_progress,
+        std::vector<bool>                                     * rules_may_be_empty);
```
Comment (HanClinto) [inline llama.cpp:13182]: Don't need this definition here, since the function is defined right below it.

**Comment by HanClinto** [comment]
> Actually, I might leave it as is for now if that looks good to you. I have some ideas about interface changes but I think that'd be a larger change that's decoupled from this one. As of now, the changes here are self-contained and strictly improvements to the status quo, in that the thrown exception here is what would have been a stack overflow or OOM in the status quo.

Yes, I think that's a fair way to approach it. Especially with the server (that can accept dynamic grammars as part of the query request), I don't think that a bad grammar should be able to crash the server -- it should probably gracefully return an error.

That said, such functionality can be added later, and as you say -- everything here is an incremental improvement from the status quo.

**Review by HanClinto — APPROVED** [review]
Looks good to me!

**Comment by mofosyne** [comment]
@nuchi added tests/test-grammar-integration.cpp and the CI approved. @HanClinto and @nuchi both agrees that this PR is good enough as a starting point for further incremental improvements.

At a glance the reviewed code does not have any obvious issues and there is already back and forth reviews. Safe to merge.

**Comment by mofosyne** [comment]
@nuchi welp, looks like 'CI / android-build (push) Failing after 1m' is still failing in the main branch. See if you can trace down the issue and make a new PR for that.

**Comment by mofosyne** [comment]
Actually scratch that. The commit ahead showed that this particular error did not happen again.

It's a timeout so must be a fluke. Wonder if I can just try kicking off that test again.
